### PR TITLE
Feat intnorm

### PR DESCRIPTION
**Pull request by bernardo-marques**
1. Added option to add preprocessing before sampling.
2. Added intensity normalisation as the only option for preprocessing with ease of extension
3. Added the necessary parameters for the configuration.

### CODE REVIEW

**Comment by bernardo-marques** [comment]
Needed changes before merge:
- inputs for booleans as booleans
- int-norm params as dictionary

**Comment by Kamnitsask** [comment]
Also needs addition of intnorm in testing

File: deepmedic/frontEnd/configParsing/trainConfig.py [hunk 1]
```diff
@@ -6,70 +6,76 @@
...
     
-    #Optional but highly suggested.
+    # Optional but highly suggested.
     SESSION_NAME = "sessionName"
-    #[REQUIRED]
+    # [REQUIRED]
     FOLDER_OUTP = "folderForOutput"
     SAVED_MODEL = "cnnModelFilePath"
-    TENSORBOARD_LOG = "tensorboardLog"
+    TENSORBOARD_LOG = "tensorboardLogtensorboardLog"
```
Comment (Kamnitsask) [inline deepmedic/frontEnd/configParsing/trainConfig.py]: something wrong happened here.

```suggestion
    TENSORBOARD_LOG = "tensorboardLog"
```

File: deepmedic/frontEnd/configParsing/trainConfig.py [hunk 2]
```diff
@@ -6,70 +6,76 @@
...
+    # ~~~~~ Training cycle ~~~~~~~~
     NUM_EPOCHS = "numberOfEpochs"
     NUM_SUBEP = "numberOfSubepochs"
     NUM_CASES_LOADED_PERSUB = "numOfCasesLoadedPerSubepoch"
     NUM_TR_SEGMS_LOADED_PERSUB = "numberTrainingSegmentsLoadedOnGpuPerSubep"
     BATCHSIZE_TR = "batchsize_train"
     NUM_OF_PROC_SAMPL = "num_processes_sampling"
+
+    # ~~~~~ Preprocessing ~~~~~~~~
+    NORM = 'norm'
```
Comment (Kamnitsask) [inline deepmedic/frontEnd/configParsing/trainConfig.py]: reminder for dictionary

File: deepmedic/frontEnd/trainSession.py [hunk 3]
```diff
@@ -195,7 +195,7 @@ def run_session(self, *args):
             self._log.print3("=======================================================\n")
 
             do_training(*([sessionTf, saver_all, cnn3d, trainer] + self._params.get_args_for_train_routine()),
-                        tensorboard_loggers=tensorboard_loggers)
+                        tensorboard_loggers=tensorboard_loggers, norm_params=self._params.get_norm_params())
```
Comment (Kamnitsask) [inline deepmedic/frontEnd/trainSession.py]: I dont think they should be separate from the rest parameters returned by  .get_args_for_train_routine().

Put them to be returned in there.

File: deepmedic/frontEnd/configParsing/trainSessionParams.py [hunk 4]
```diff
@@ -455,6 +468,9 @@ def get_path_to_load_model_from(self):
     def get_tensorboard_bool(self):
         return self.tensorboardLog
 
+    def get_norm_params(self):
```
Comment (Kamnitsask) [inline deepmedic/frontEnd/configParsing/trainSessionParams.py]: remove. Return the parameters together with the standard arguments, under the "Pre-processing" section

File: deepmedic/frontEnd/configParsing/trainSessionParams.py [hunk 5]
```diff
@@ -324,6 +324,19 @@ def __init__(self,
         # Preprocessing
```
Comment (Kamnitsask) [inline deepmedic/frontEnd/configParsing/trainSessionParams.py]:
```suggestion
        # ==== Preprocessing =====
        # Padding
```

File: deepmedic/frontEnd/configParsing/trainSessionParams.py [hunk 6]
```diff
@@ -324,6 +324,19 @@ def __init__(self,
...
         self.pad_input_imgs = cfg[cfg.PAD_INPUT] if cfg[cfg.PAD_INPUT] is not None else True
 
+        # Normalisation
+        self.norm = cfg[cfg.NORM]
```
Comment (Kamnitsask) [inline deepmedic/frontEnd/configParsing/trainSessionParams.py]: We should add a comment saying what values are expected? At least if it's not in the config file itself.

File: deepmedic/frontEnd/configParsing/trainConfig.py (hunk 2 above)
Comment (Kamnitsask) [inline deepmedic/frontEnd/configParsing/trainConfig.py]:
```suggestion
    NORM_ZSCORE = 'norm_zscore'
```

File: deepmedic/frontEnd/configParsing/trainSessionParams.py [hunk 7]
```diff
@@ -324,6 +324,19 @@ def __init__(self,
...
+        if self.norm:
```
Comment (Kamnitsask) [inline deepmedic/frontEnd/configParsing/trainSessionParams.py]:
```suggestion
        if self.norm_zscore:
```

File: deepmedic/frontEnd/configParsing/trainSessionParams.py [hunk 8]
```diff
@@ -324,6 +324,19 @@ def __init__(self,
...
+            self.norm_params = {'int_norm': cfg[cfg.INT_NORM],
```
Comment (Kamnitsask) [inline deepmedic/frontEnd/configParsing/trainSessionParams.py]:
```suggestion
            self.norm_zscore_params = {'int_norm': cfg[cfg.INT_NORM],
```

File: deepmedic/frontEnd/configParsing/trainSessionParams.py [hunk 9]
```diff
@@ -586,6 +602,14 @@ def print_params(self):
         logPrint("Check whether input data has correct format (can slow down process) = " + str(self.run_input_checks))
         logPrint("~~Pre Processing~~")
         logPrint("Pad Input Images = " + str(self.pad_input_imgs))
+        logPrint("~~Normalization~~")
+        logPrint("Normalisation = " + str(bool(self.norm)))
```
Comment (Kamnitsask) [inline deepmedic/frontEnd/configParsing/trainSessionParams.py]: zscore

File: deepmedic/routines/training.py [hunk 10]
```diff
@@ -187,7 +187,9 @@ def do_training(sessionTf,
                 run_input_checks,
 
                 # ------- loggers -------
-                tensorboard_loggers=None
+                tensorboard_loggers=None,
+                # ---- Normalization ----
+                norm_params=None
```
Comment (Kamnitsask) [inline deepmedic/routines/training.py]: I would kindly ask us to remove the defaults from here.
You want to be defining the defaults only in trainSessionParams.py. Defaults concentrated there in one place.
And the code to throw an error if for any reason your SessionParams.py forgets to return a value (bug)

```suggestion
                norm_params
```

Also, move it further up, to the "pre-processing " section

File: deepmedic/routines/training.py [hunk 11]
```diff
@@ -187,7 +187,9 @@ def do_training(sessionTf,
                 run_input_checks,
 
                 # ------- loggers -------
-                tensorboard_loggers=None
+                tensorboard_loggers=None,
```
Comment (Kamnitsask) [inline deepmedic/routines/training.py]: I would remove the default after merging the dev

File: deepmedic/dataManagement/sampling.py [hunk 12]
```diff
@@ -34,183 +37,214 @@
...
+
+    id_str = "[SAMPLER-TR|PID:" + str(os.getpid()) + "]" if train_or_val == "train" \
+        else "[SAMPLER-VAL|PID:" + str(os.getpid()) + "]"
     start_time_sampling = time.time()
     training_or_validation_str = "Training" if train_or_val == "train" else "Validation"
-    
-    log.print3(id_str+" :=:=:=:=:=:=: Starting to sample for next [" + training_or_validation_str + "]... :=:=:=:=:=:=:")
-    
+
+    log.print3(id_str + " :=:=:=:=:=:=: Starting to sample for next [" +
```
Comment (Kamnitsask) [inline deepmedic/dataManagement/sampling.py]: shorten to put in 1 line (I ll rename the  variable)

File: deepmedic/dataManagement/sampling.py [hunk 13]
```diff
@@ -34,183 +37,214 @@
...
+                log.print3(traceback.format_exc())
                 worker_pool.terminate()
-                worker_pool.join() # Will wait. A KeybInt will kill this (py3)
+                worker_pool.join()  # Will wait. A KeybInt will kill this (py3)
                 raise e
-            except: # Catches everything, even a sys.exit(1) exception.
-                log.print3(id_str+"\n\n ERROR: Unexpected error in getSampledDataAndLabelsForSubepoch(). System info: ", sys.exc_info()[0])
+            except:  # Catches everything, even a sys.exit(1) exception.
+                log.print3(
+                    id_str + "\n\n ERROR: Unexpected error in getSampledDataAndLabelsForSubepoch(). System info: ",
```
Comment (Kamnitsask) [inline deepmedic/dataManagement/sampling.py]: break? Shorten?

File: deepmedic/dataManagement/preprocessing.py [hunk 14]
```diff
@@ -10,58 +10,195 @@
...
+
     return unpaddedArray1
 
-    
 
+def neg_val_check(img, log):
+    is_neg_int = img < 0
+    num_voxels_neg_int = np.sum(is_neg_int)
+    if num_voxels_neg_int > 0:
+        log.print3("[WARNING] One image has " + str(num_voxels_neg_int) +
```
Comment (Kamnitsask) [inline deepmedic/dataManagement/preprocessing.py]:
```suggestion
        log.print3("WARN: One image has " + str(num_voxels_neg_int) +
```

File: deepmedic/dataManagement/preprocessing.py [hunk 15]
```diff
@@ -10,58 +10,195 @@
...
+    if norm_params['cutoff_std']:
+        cutoff_types += ['Standard Deviation']
+    if norm_params['cutoff_mean']:
+        cutoff_types += ['Whole Image Mean']
+
+    log.print3(id_str + "Normalising " + str(num_channels) + " channel(s) with the following cutoff type(s): " +
+               ', '.join(list(cutoff_types)) if cutoff_types else 'None')
+
+
+def normalize_images(log, channels, roi_mask, norm_params=None, id_str='', verbose=False):
```
Comment (Kamnitsask) [inline deepmedic/dataManagement/preprocessing.py]: rename to something with "zscore"

File: deepmedic/dataManagement/preprocessing.py [hunk 16]
```diff
@@ -10,58 +10,195 @@
...
+def neg_val_check(img, log):
+    is_neg_int = img < 0
+    num_voxels_neg_int = np.sum(is_neg_int)
+    if num_voxels_neg_int > 0:
+        log.print3("[WARNING] One image has " + str(num_voxels_neg_int) +
+                   " voxels with intensity below zero! Unexpected!\n" +
+                   "The voxels with negative ints have (min, max, mean) = (" + str(np.min(img[is_neg_int])) + ", " +
+                   str(np.max(img[is_neg_int])) + ", " + str(np.mean(img[is_neg_int])) + ").")
+
+
```
Comment (Kamnitsask) [inline deepmedic/dataManagement/preprocessing.py]: Likely we should make a Pre-Processor (or normalizer?) class and Z-score-norm should be a child?

Comment (bernardo-marques) [reply deepmedic/dataManagement/preprocessing.py]: agree

File: deepmedic/dataManagement/sampling.py (hunk 13 above)
Comment (bernardo-marques) [reply deepmedic/dataManagement/sampling.py]: Not sure I know what you mean

**Pull request merged**